KAFKA-14912:Add a dynamic config for remote index cache size

hudeqi opened the pull request:
As https://issues.apache.org/jira/browse/KAFKA-14912 described, add a dynamic config for remote index cache size.

Commit:
KAFKA-14912:Add a dynamic config for remote index cache size

Commit:
fix conlict

Commit:
fix

Commit:
fix conflict

Commit:
use Caffeine weigher

Commit:
Merge branch 'trunk' of https://github.com/hudeqi/kafka into hdq_remote_index_cache

Commit:
fix unit test

Commit:
fix

Commit:
remove unused import

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix ut

Commit:
Merge branch 'trunk' of https://github.com/hudeqi/kafka into hdq_remote_index_cache

Commit:
fix conflict

Commit:
minor fix

hudeqi commented:
Hi, the previous [PR](https://github.com/apache/kafka/pull/14243) was deleted and closed by mistake. :) Sorry for the hard work, please review again, thank you. @divijvaidya @showuon @kamalcph @satishd

hudeqi commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -112,27 +115,55 @@ public class RemoteIndexCache implements Closeable {
+    public void resizeCacheSize(long remoteLogIndexFileCacheSize) {
+        lock.writeLock().lock();
+        try {
+            // When resizing the cache, we always start with an empty cache. There are two main reasons:
+            // 1. Resizing the cache is not a high-frequency operation, and there is no need to fill the data in the old
+            // cache to the new cache in time when resizing inside.
+            // 2. Since the eviction of the caffeine cache is cleared asynchronously, it is possible that after the entry
+            // in the old cache is filled in the new cache, the old cache will clear the entry, and the data in the two caches
+            // will be inconsistent.
+            internalCache.invalidateAll();
+            log.info("Invalidated all entries in the cache and triggered the cleaning of all index files in the cache dir.");
+            internalCache = initEmptyCache(remoteLogIndexFileCacheSize);
```
https://github.com/apache/kafka/pull/14243#discussion_r1320630057

satishd replied:
Do we have a JIRA tracking the improvement suggested in the comment?

showuon replied:
Good point!
Created https://issues.apache.org/jira/browse/KAFKA-15536 for this issue, and assigned to @hudeqi . Thanks.

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -79,6 +79,7 @@ public class RemoteIndexCache implements Closeable {
     private static final String TMP_FILE_SUFFIX = ".tmp";
     public static final String REMOTE_LOG_INDEX_CACHE_CLEANER_THREAD = "remote-log-index-cleaner";
     public static final String DIR_NAME = "remote-log-index-cache";
+    private static final long DEFAULT_REMOTE_INDEX_CACHE_SIZE_BYTES = 1024 * 1024L;
```
Let's move this constant to `RemoteIndexCacheTest` since this is used only in the test.

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -112,27 +115,55 @@ public class RemoteIndexCache implements Closeable {
+    private Cache<Uuid, Entry> internalCache;
 
+    // Visible for testing
     public RemoteIndexCache(RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
```
We can remove this constructor and use the other constructor from the test.

kamalcph reviewed (approved):
LGTM!

Commit:
minor fix

showuon commented:
@hudeqi , are you sure this branch is 100% the same as the one in https://github.com/apache/kafka/pull/14243 since you said you've deleted that branch?

showuon commented:
Hmm... alright, let's focus on this PR, then. 

![image](https://github.com/apache/kafka/assets/43372967/085019ab-5cce-41bb-8fb0-aa902b40cecc)

hudeqi commented:
> Hmm... alright, let's focus on this PR, then.
> 
> ![image](https://user-images.githubusercontent.com/43372967/267619603-085019ab-5cce-41bb-8fb0-aa902b40cecc.png)

I tried, but can not, may be because: I delete the origin branch, and then the old pr is closed automatically.

hudeqi commented:
This PR is guaranteed to be the same as the old one, because I deleted the branch by mistake, then pushed it to the remote, and then raised a new PR.

showuon commented:
> This PR is guaranteed to be the same as the old one, because I deleted the branch by mistake, then pushed it to the remote, and then raised a new PR.

Thanks for the confirmation. Yes, I also checked the commit hash, everything is the same as the old one till this commit: [4c41e98](https://github.com/apache/kafka/pull/14381/commits/4c41e983a49d885d2d4ab106e0dd5551c6dc5768). I reviewed the latest 2 commits and LGTM! Thanks for the work. Let's wait for the CI build results:
https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14381/3/

Thanks.

showuon commented:
Failed tests are unrelated:
```
    Build / JDK 8 and Scala 2.12 / kafka.network.SocketServerTest.testSaslReauthenticationFailureNoKip152SaslAuthenticate()
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBootstrapZkMigrationRecord()
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBootstrapZkMigrationRecord()
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.NamedTopologyIntegrationTest.shouldRemoveOneNamedTopologyWhileAnotherContinuesProcessing()
    Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testOffsetTranslationBehindReplicationFlow()
    Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector
    Build / JDK 20 and Scala 2.13 / kafka.api.SslAdminIntegrationTest.testSynchronousAuthorizerAclUpdatesBlockRequestThreads()
    Build / JDK 20 and Scala 2.13 / kafka.api.UserQuotaTest.testThrottledProducerConsumer(String).quorum=kraft
    Build / JDK 20 and Scala 2.13 / kafka.network.DynamicConnectionQuotaTest.testDynamicListenerConnectionCreationRateQuota()
    Build / JDK 20 and Scala 2.13 / kafka.server.DescribeClusterRequestTest.testDescribeClusterRequestIncludingClusterAuthorizedOperations(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector
    Build / JDK 11 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testCreateTokenForOtherUserFails(String).quorum=kraft
    Build / JDK 11 and Scala 2.13 / kafka.api.TransactionsTest.testBumpTransactionalEpoch(String).quorum=kraft
```

showuon commented:
@satishd , not sure if you want to have another look. If no, I'll merge it. Thanks.

hudeqi commented:
Do you plan to combine this PR to 3.6? That is, can it be merged now? @satishd @showuon

satishd commented:
@hudeqi We do not plan to merge this into 3.6.0 release as we have been merging only release blockers for more than a week. 

@showuon I will take a look at this PR next week.

divijvaidya commented:
@satishd (gentle reminder), we are waiting for your review on this one.

satishd commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1129,3 +1131,50 @@ class DynamicProducerStateManagerConfig(val producerStateManagerConfig: Producer
+  override def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    newConfig.values.forEach { (k, v) =>
+      if (reconfigurableConfigs.contains(k)) {
+        val newValue = v.asInstanceOf[Long]
```
All the values are assumed to be of type `long`. Please have better checks based on the respective property names and values.

hudeqi replied:
updated.

satishd commented on storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java:
```diff
@@ -197,7 +197,7 @@ public final class RemoteLogManagerConfig {
                                   LOW,
                                   REMOTE_LOG_METADATA_CUSTOM_METADATA_MAX_BYTES_DOC)
-                  .defineInternal(REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES_PROP,
+                  .define(REMOTE_LOG_INDEX_FILE_CACHE_TOTAL_SIZE_BYTES_PROP,
```
Filed https://issues.apache.org/jira/browse/KAFKA-15535 to followup on adding documentation.

hudeqi replied:
Thanks, I have assigned it to myself.

hudeqi replied:
This configuration name and the corresponding doc have been provided by you before, but have not been used. This time, the configuration name is directly reused.
I also read other configuration docs related to tiered storage and found nothing missing. @satishd

satishd reviewed (commented):
Thanks @hudeqi for the PR. Overall LGTM, left a couple of minor comments.

Commit:
fix conflict

Commit:
minor fix

Commit:
fix this-escape warning

satishd commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -1140,12 +1140,14 @@ class DynamicRemoteLogConfig(server: KafkaBroker) extends BrokerReconfigurable w
+          val newValue = v.asInstanceOf[Long]
+          val oldValue = getValue(server.config, k)
+          if (newValue != oldValue) {
+            val errorMsg = s"Dynamic remote log manager config update validation failed for $k=$v"
```
Have this defined in the below if condition as it is used only for that. 

```
if (newValue <= 0) {
  val errorMsg = s"Dynamic remote log manager config update validation failed for $k=$v"
  throw new ConfigException(s"$errorMsg, value should be at least 1")
}
```

hudeqi replied:
Updated. Thanks.

satishd reviewed (commented):
Thanks @hudeqi for the updates. Left a minor comment.

Commit:
fix

satishd reviewed (approved):
Thanks @hudeqi for addressing the review comments. LGTM.

hudeqi commented:
> Thanks @hudeqi for addressing the review comments. LGTM.

Thanks your review, @satishd please merge it. I have resolved several conflicts. I see that there seems to be another PR that conflicts with this PR :)

satishd commented:
There are a few unrelated test failures, merging it to trunk.
